### PR TITLE
Issue #10688 add Jetty 12.0.x ee8 osgi

### LINKED ISSUE #10688: Introduce Jetty 12 ee8 osgi layer
**Jetty version(s)**
12

**Jetty Environment**
ee8

**Java version/vendor** `(use: java -version)`
Any

**OS type/version**
Any

**Description**
As pointed out in a eclipse releng issue https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/pull/1430#issuecomment-1754416750

The current release of Jetty 12.0.2 does not contain a ee8 OSGI layer.

There does exist an OSGi layer for ee10 and ee9 though.

* https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/osgi/
* https://repo1.maven.org/maven2/org/eclipse/jetty/ee9/osgi/

Looks like the jetty-ee8/osgi layer isn't even present in the source tree.

https://github.com/eclipse/jetty.project/tree/jetty-12.0.x/jetty-ee8

### PR DESCRIPTION
**Pull request by olamy**
issue #10688

### CODE REVIEW

File: jetty-ee8/pom.xml [hunk 1]
```diff
@@ -46,9 +50,10 @@
     <javax.mail.glassfish.version>1.4.1.v201005082020</javax.mail.glassfish.version>
     <jetty.servlet.api.version>4.0.6</jetty.servlet.api.version>
     <jsp.impl.version>9.0.52</jsp.impl.version>
-    <modify-sources-plugin.version>1.0.6</modify-sources-plugin.version>
+    <modify-sources-plugin.version>1.0.7-SNAPSHOT</modify-sources-plugin.version>
```
Comment (olamy) [inline jetty-ee8/pom.xml]: this will need a release of this plugin before we merge that

Comment (merks) [reply jetty-ee8/pom.xml]: @olamy 

You're awesome for taking such rapid action!  🏅

Comment (olamy) [reply jetty-ee8/pom.xml]: release done. just waiting sync to central repo.

Comment (olamy) [reply jetty-ee8/pom.xml]: update done

**Comment by claesrosell** [comment]
I am playing with this now and I found this problem so far:
The Bundle-Activator header seems to be wrong in the generated MANIFEST.MF of the bundle "org.eclipse.jetty.ee8.osgi.boot". 

The class of the Activator is wrongly named "EE9Activator" and the generated header is: `org.eclipse.jetty.ee8.osgi.boot.ee8Activator` with lowercase EE.

**Comment by claesrosell** [comment]
BTW, I didn't intend to do a complete review, I just happened to press the wrong button when I intended to add some ee8/ee9 comments. Sorry.

**Comment by olamy** [comment]
> BTW, I didn't intend to do a complete review, I just happened to press the wrong button when I intended to add some ee8/ee9 comments. Sorry.

no worries. we are happy to get feedback from real users :) 
ah you found a problem in the tool we use for translating sources from ee9 to ee8.
I need to fix the generated class name.

**Comment by olamy** [comment]
back to draft as I found some issues

**Comment by claesrosell** [comment]
> back to draft as I found some issues

Should I await modify-sources-plugin 1.0.8 before testing this?

**Comment by olamy** [comment]
> > back to draft as I found some issues
> 
> Should I await modify-sources-plugin 1.0.8 before testing this?

yes sorry for that. But this has trigger some more issues :(

**Comment by olamy** [comment]
@claesrosell problem fixed. feel free to test again. Thanks for reporting the issue!

**Review by merks — APPROVED** [review]
Sorry, but I don't have the knowledge or skills to actually review the changes.

Only @claesrosell can really confirm if the generated results are correct and functional when used by BIRT...

**Comment by claesrosell** [comment]
> Sorry, but I don't have the knowledge or skills to actually review the changes.
> 
> Only @claesrosell can really confirm if the generated results are correct and functional when used by BIRT...

I will try to test the latest changes and see if I can get the BIRT WebViewer to work.

**Comment by olamy** [comment]
@merks @claesrosell thanks for your help! 
No rush next release will be in 3 weeks ish

**Comment by janbartel** [comment]
@olamy I don't  see any ee8.osgi tests on the CI build for this PR?

**Comment by olamy** [comment]
> @olamy I don't see any ee8.osgi tests on the CI build for this PR?

@janbartel 
yup forgot that. This obviously needs more fix...

the sources doesn't manage such code `res.add(mavenBundle().groupId("org.eclipse.jetty.toolchain").artifactId("jetty-javax-servlet-api").versionAsInProject().start());`

I need to figure out something different.

**Comment by claesrosell** [comment]
Hi again. I was able to look into this again and with these latest changes I come a bit further in our use-case. However when the EE8Activator.ServerTracker picks up our server it fails with an StringIndexOutOfBoundsException
```
java.lang.StringIndexOutOfBoundsException: begin 0, end -1, length 0
	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:4606)
	at java.base/java.lang.String.substring(String.java:2709)
	at org.eclipse.jetty.ee8.osgi.boot.EE8Activator$ServerTracker.addingService(EE8Activator.java:129)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:947)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:875)
```
It is these lines that causes the exception:
```
126            if (contributedBundles != null) {
127                StringBuffer strbuff = new StringBuffer();
128                contributedBundles.stream().forEach(b -> strbuff.append(b.getSymbolicName()).append("|"));
129 -->          containerScanBundlePattern = strbuff.toString().substring(0, strbuff.length() - 1);
130           }
```
contributedBundles is not null, but an empty list: This causes the strbuff to be empty ( strbuff.length is 0 ) which leads to the exception since the end position ends up being -1.

**Comment by olamy** [comment]
@claesrosell thanks for the report. I have fixed. It looks we have some issues with paxexam and logging so we couldn't see this issue. 
we are still looking at the reason why our tests are still not working but we made some progress :)

**Comment by olamy** [comment]
@merks @claesrosell we plan to merge this soon have a release probably next week

**Comment by claesrosell** [comment]
I will work on getting the Birt webviewer working today. The problem as it looks now is that the Birt Webviewer creates a WebAppContext and register it as an OSGI service and that scenario does not seem to be supported by Jetty anymore. So I will probably need to create a web-bundle instead.

**Review by joakime — APPROVED** [review]

**Pull request merged**